### DISCUSSION TITLE
The on function

### DISCUSSION THREAD

**Discussion by lukhomdingi** [ngrx/platform#3104] [Q&A] [answered]
Hello,

Could you provide a proper way of using the on function with the following syntax:

```
// I'd like to understand the second type better.
export const loginSuccessHandler = on<{user: User}, any>(AuthApiActions.loginSuccess, (state, { user }) => ({ ...state, user }));
```

**Comment by markostanimirovic** [comment, marked as answer]
@lukhomdingi 

`on` function is intended for use within the `createReducer` function. Then the type of action creators will be automatically inferred.

However, if you want to use `on` function separately, you can manually declare the type of action creator array:

```ts
export const onLoginSuccess = on<fromAuth.State, [typeof AuthApiActions.loginSuccess]>(
  AuthApiActions.loginSuccess,
  (state, { user }) => ({ ...state, user })
);
```
